### PR TITLE
Sync with obelisk/develop

### PR DESCRIPTION
**Pull request by romefeller**
Sync cli-extras with obelisk (develop branch) in order to refactor obelisk-cliapp

### CODE REVIEW

File: src/Cli/Extras/Process.hs [hunk 1]
```diff
@@ -132,10 +131,10 @@ readCreateProcessWithExitCode procSpec = do
 -- | Like `System.Process.readProcess` but logs the combined output (stdout and stderr)
 -- with the corresponding severity.
```
Comment (plt-amy) [inline src/Cli/Extras/Process.hs:132]: Document what each field of the tuple corresponds to

File: src/Cli/Extras/Process.hs [hunk 2]
```diff
@@ -214,18 +210,30 @@ callCommand cmd = do
...
+    )
+    (liftIO . Process.cleanupProcess)
+    (\case
+      (_, Just out, Just err, p) -> do
+        f out err
+        (out, err) <$ (exitCodeToException process =<< waitForProcess p)
+      _ -> error "withProcess: createProcess did not provide handles for CreatePipe as expected"
+    )
+
+-- | Runs a process to completion failing if it does not exit cleanly.
```
Comment (plt-amy) [inline src/Cli/Extras/Process.hs:228]: Failing how?

File: src/Cli/Extras/Types.hs [hunk 3]
```diff
@@ -37,9 +41,42 @@ type CliLog m = MonadLog Output m
 
 type CliThrow e m = MonadError e m
 
+-- | Log a message to the console.
+--
+-- Logs safely even if there are ongoing spinners.
```
Comment (plt-amy) [inline src/Cli/Extras/Types.hs:46]: What does "safely" mean here?

File: src/Cli/Extras/Logging.hs [hunk 4]
```diff
@@ -21,77 +20,47 @@ module Cli.Extras.Logging
...
-
-      -- Respect the user’s TERM environment variable. Dumb terminals
-      -- like Eshell cannot handle lots of control sequences that the
-      -- spinner uses.
-      termEnv <- lookupEnv "TERM"
-      let isDumb = termEnv == Just "dumb"
-
-      return $ isTerm && not inShellCompletion && not isDumb
-
 newCliConfig
```
Comment (plt-amy) [inline src/Cli/Extras/Logging.hs:57]: Haddock?

File: src/Cli/Extras/Process.hs [hunk 5]
```diff
@@ -244,14 +252,20 @@ streamToLog stream = fix $ \loop -> do
 waitForProcess :: MonadIO m => ProcessHandle -> m ExitCode
 waitForProcess = liftIO . Process.waitForProcess
 
+-- | Converts an 'ExitCode' to an exception when it's non-zero.
```
Comment (plt-amy) [inline src/Cli/Extras/Process.hs:255]: Not too convinced about the naming/documentation here: It doesn't _convert_ an `ExitCode` to an exception, it _throws_ an exception on `ExitFailure`. Can we change the name/doc to better reflect this?

**Review by plt-amy — CHANGES_REQUESTED** [review]

**Comment by plt-amy** [comment]
Build failures are my bad, I only had ghcid running with 8.8.4

File: src/Cli/Extras/Process.hs [hunk 6]
```diff
@@ -61,6 +62,10 @@ import Control.Monad.Log (Severity (..))
 import Cli.Extras.Logging (putLog, putLogRaw)
 import Cli.Extras.Types (CliLog, CliThrow)
 
+#if !(MIN_VERSION_base(4, 13, 0))
+import Control.Monad.Fail (MonadFail)
+#endif
```
Comment (plt-amy) [inline src/Cli/Extras/Process.hs:67]: @madeline-os should I use these CPP guards to only import MonadFail when needed, should I disable the "redundant imports" warning, or should I just leave the warning there?

File: src/Cli/Extras/Types.hs (hunk 3 above)
Comment (plt-amy) [reply src/Cli/Extras/Types.hs:50]: 👍

File: src/Cli/Extras/Process.hs (hunk 6 above)
Comment (madeline-os) [reply src/Cli/Extras/Process.hs:67]: Go for CPP guards, or just leave the warning in there if it already exists.

**Review by madeline-os — APPROVED** [review]

**Review by plt-amy — APPROVED** [review]

**Pull request merged**